README.md
dkim_policy
===========

## Query DKIM signing policies of a domain


### Native Solution for Exim

####(Added on 2016/01/11)
Although the dkim_policy Perl script is useful for manual testing, and 
has some advantages for the use in Exim, I finally managed to replicate 
very similar functionality with using just native Exim commands (namely
the dnsdb() directive). For most Exim administrators, the native version
of DKIM policy checking is more acceptable. Please find the native 
solution in the file EXIM.conf in this package.

### Perl Script

This script was written to allow better exploiting the possibilities of 
DKIM in Exim.

In the version 4.88 of Exim, available in the time of writing this script,
EXIM supports the DKIM standard for email authentication, which is very
helpful in preventing spam, forgery, and abuse. However, the potential of
DKIM is not fully exploited in Exim, because it only tests DKIM signatures
present in the respective email message. It does not query the DNS at
every sender to try finding out whether the domain owner requires all
messages to be signed or not. It means, in stock Exim (as far as I could
find) you currently cannot reject or quarantine unsigned incoming messages
that are supposed to be signed.

The tool checks if any of the following three policies exists in the
DNS records of given domain and returns the value 0 (success) in such case.
When no signing policy is found, it returns 1 (error).

 - RFC4870(historical) DomainKeys sender signing policies
 - early draft DKIM sender signing policies
 - Author Domain Signing Practices (ADSP)

The Perl script uses the CPAN module Mail::DKIM and its subclasses
Mail::DKIM::Verifier and Mail::DKIM::Policy. It was written in Perl5.24,
using Mail-DKIM-0.40, on a FreeBSD machine. Slight modifications may be
needed on other platforms (for example you may need to adjust the path
to Perl on the first line of the script).

This utility is just a quick intermittent solution allowing the rejection
of unsigned messages. I suppose that later version of Exim will have
such functionality built in. However, the script may be useful also for
other purposes, including quick manual or automated testing of DNS DKIM
settings.

Administrators with busy servers should be aware that calling the script
at every email will add processing time and bandwidth usage for the
additional DNS queries.


Installation:
-------------

Unzip the distribution package and place the dkim_policy.pl to some
location on your server (avoid publicly accessible folders such as the www
tree), and turn on the execution permission bit on the file. On some platforms
you may need to change the path to Perl on the first line of the script. In
doubts about the right path, consult other Perl scripts on your machine.


Required:
- [Perl 5](http://www.perl.org/)
- Perl module [Mail::DKIM](http://search.cpan.org/~jaslong/Mail-DKIM-0.40/lib/Mail/DKIM.pm)
- Perl module [Getopt::Long](http://search.cpan.org/~jv/Getopt-Long-2.49.1/lib/Getopt/Long.pm)

In case the modules are not installed on your system, you can add them in
the following way from command line:
```
 cpan install Mail::DKIM
 cpan install Getopt::Long
```


Examples of use:
----------------

The script prints out the following usage information when invoked with:

`./dkim_policy.pl -h`

```
dkim_policy.pl  DKIM Policy Checker v1.02 (2017-01-10)
  DNS Query of DKIM signing policies (DK/DKIM/ADSP) of a given domain
  Written by Ivo Truxa (c) 2017 <truxa@truxoft.com>

  usage: dkim-policy.pl [options...] FQDN
     Otions:
        -d, --details     display a detailed verbose output
        -c, --changelog   display the version history and exit
        -v, --version     display the version number and exit
        -h, --help        showing this information

     FQDN                 Fully Qualified Domain Name to be tested

  In the default simple mode, the script returns 'all' and the return code 0 (success)
  when any of DK/DKIM/ADSP policies require all messages to be signed. In all other
  cases (no policy defined, or allowing some messages without signatures), the return
  code is 1 (error), and the return string is 'some/undef'. In the detailed mode, the
  script returns DK/DKIM/ADSP policies as found in the DNS.
```

You can use the tool for polling signing policies manually in the following way:

`./dkim_policy.pl -d somedomain.com`

It may be useful also for other purposes, but originally it was written to help 
Exim rejecting unsigned email from domains that declare the use of DKIM on all messages. 
Below, you can find a sample section of an Exim configuration file, demonstrating the 
use of the built-in DKIM function together with the dkim_policy script.

First of all, we have to enable the DKIM ACL for every message, regardless whether 
it is signed or not. We can achieve it with the following global parameter at the 
beginning of the Exim configuration file:

`dkim_verify_signers = $sender_address_domain`

Personally, I only base the reject decision on senders' DKIM, so testing 
$sender_address_domain is all I need. If you want to test all DKIM signatures 
of every message (in case of multiple DKIM's in a message), you need to append 
':$dkim_singers'.

Then, you also need to disable the DKIM control for authorized users, trusted 
relays, and for some other cases, with 'control = dkim_disable_verify'. This 
is well covered in most Exim DKIM Howto's.

The DKIM ACL block below works with some previously defined domain lists. 
Namely the $dkim_domains list (for important well known and often abused 
domains, like those of Gmail, Paypal, many banks, etc.), and $local_domains, 
where I store all domain names served by my server. It is up to you whether 
you populate the list from a flat file with lsearch, use a database, or fill 
them manually.

The configuration section is well commented, so read through it, remove or 
add comments, and adjust it to suit your needs. The use of the dkim_policy 
script is demonstrated at the bottom of the section, where it helps denying 
unsigned messages from senders who's DNS records claim all messages to be 
signed.

In the case the script call fails, the shell returns error (1), and Exim will
process the message as if the policy was undefined (no DKIM enforcement).

```
# ----------- ACL DKIM ----------------------------------------------
acl_check_dkim:
# Ivo Truxa 2017-01-09
# logwrites mostly for testing and debugging purposes, remove later

  # Just setting a frequently used debugging logwrite text
  warn  set acl_m_dklog = $dkim_verify_status, D=$sender_address_domain, SG=$dkim_domain, KT=$dkim_key_testing, SUB=$dkim_key_nosubdomains, S=$dkim_selector, R=$dkim_verify_reason

  # We accept failed signatures too, when DKIM is set to the test mode in sender's DNS
  accept condition      = $dkim_key_testing
        add_header      = X-DKIM: TEST mode result for $sender_address_domain (signer=$dkim_cur_signer): $dkim_verify_status; $dkim_verify_reason
#       logwrite        = DKIM DEBUG 01: TEST=$acl_m_dklog

  # Accept valid/passed signatures
  accept dkim_status    = pass
        add_header      = X-DKIM: $dkim_verify_status (address=$sender_address domain=$dkim_cur_signer), signature is good.
#       logwrite        = DKIM DEBUG 02: OK=$acl_m_dklog

  # Deny failures (logging only, while testing; enable later by uncommenting 'deny' and removing the verb 'warn')
  #deny message         = DKIM test failed: $dkim_verify_reason. Please use only authorized email address & SMTP server!
  warn  dkim_status     = fail
        add_header      = X-DKIM: $dkim_cur_signer ($dkim_verify_status); $dkim_verify_reason
#       logwrite        = DKIM DEBUG 10: ERR=$acl_m_dklog

  # Deny invalid signatures
  #deny message         = DKIM signature could not be verified: $dkim_verify_reason. Please review the DNS records of $sender_address_domain
  # For the moment we are accepting invalid messages (owner's fault, no evidence of a forgery)
  # if denying needed later, uncomment 'deny' above, and remove the 'warn' verb below
  warn  dkim_status     = invalid
        add_header      = X-DKIM: $dkim_cur_signer ($dkim_verify_status); $dkim_verify_reason
#       logwrite        = DKIM DEBUG 11: ERR=$acl_m_dklog

  # Deny missing signatures at important known signers, frequently used domains,
  # and at local domains (except of someexception.com,...)
  # (by doing so here, we save some excess processing time and bandwidth
  #  needed for frequent DNS TXT lookups in the next section)
  deny  message         = DKIM signature missing! The policy of the domain $sender_address_domain enforces DKIM signatures on all email.
        dkim_status     = none
        sender_domains  = +dkim_domains : +local_domains
        !sender_domains = someexception.com : other.exception.org : more.exceptions.com
        add_header      = X-DKIM: $sender_address: signature is missing.
#       logwrite        = DKIM DEBUG 12: ERR=$acl_m_dklog

  # query DNS for the DKIM signing policy, and reject message if signing enforced
  deny  message         = DKIM signature missing! The policy of the domain $sender_address_domain enforces DKIM signatures on all email.
        dkim_status     = none
        !sender_domains = +dkim_domains : +local_domains
        condition       = ${run{/path/to/dkim_policy.pl $sender_address_domain}{yes}{no}}
#       logwrite        = DKIM DEBUG 13: P=$value, D=$sender_address_domain

  # And accept anything else (i.e. senders without DKIM policy, or with a neutral one)
  accept
```


Contact:
--------

You can send your questions or kudos to [Ivo Truxa](mailto:truxa@truxoft.com)

You can also submit issues, or modifications through the
[GitHub repository](https://github.com/truxoft/dkim_policy/)


License:
--------

dkim_ is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

dkim_policy is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
dkim_policy; if not, write to the Free Software Foundation, Inc., 51 Franklin
Street, Fifth Floor, Boston, MA 02110, USA




CHANGELOG
CHANGELOG for the dkim_policy.pl script
========================================

dkim_policy.pl  DKIM Policy Checker v1.03 (2017-01-11)
  DNS Query of DKIM signing policies (DK/DKIM/ADSP) of a given domain
  Written by Ivo Truxa (c) 2017 <truxa@truxoft.com>

  1.03 [Ivo Truxa] 2017/01/11
       - added other forgotten policy strings (o=!, dkim=all, dkim=discardable)
  1.02 [Ivo Truxa] 2017/01/10
       - return code 0 (enforced) or 1 (undefined) kept, but STDOUT prints a verbose result
  1.01 [Ivo Truxa] 2017/01/09
       - inversed output - 0: success = DKIM signing enforced
       -                   1: error = policy undefined or signing not enforced
       - policy loop interrupted at the 1st hit, to save time, processing power, and bandwidth
       - full policy loop for detailed verbose mode kept
       - cosmetic formatting changes, editing of help texts
  1.00 [Ivo Truxa] 2017/01/08
       - this tool was written on FreeBSD, using Perl module Mail-DKIM-0.40 and Perl5-24
       - at this date, it was not tested on other platforms or with other versions of Perl


dkim_policy.pl
#!/usr/local/bin/perl
# dkim-policy.pl      License: GPLv2
# Ivo Truxa  (c) 2017  <truxa@truxoft.com>

# ------- VERSION -----------------------------------------------------
my $version = "1.03";
my $verdate = "2017-01-11";

# ------ PURPOSE ------------------------------------------------------
# The purpose of the script is returning of one of the available DKIM
# policies available in DNS records of a given domain (if available).
# This function is currently not availble in the built-in DKIM
# processing in Exim mail server, and it is necessary if rejecting or
# flagging of DKIM-unsigned messages is required. Before rejecting
# an unsigned message, we need to know whether the messages are
# supposed and required to be signed at all.
#
# The tool checks if any of the following three policies exists in the
# DNS records of given domain and returns the value 0 in such case.
# When no signing policy is found, it returns 1.
#  - RFC4870(historical) DomainKeys sender signing policies
#  - early draft DKIM sender signing policies
#  - Author Domain Signing Practices (ADSP)
#
# ---------------------------------------------------------------------

my $u1 = "\ndkim_policy.pl  DKIM Policy Checker v".$version." (".$verdate.")                     \n".
         "  DNS Query of DKIM signing policies (DK/DKIM/ADSP) of a given domain                  \n".
         "  Written by Ivo Truxa (c) 2017 <truxa\@truxoft.com>                                   \n";
my $u2 = "  usage: dkim-policy.pl [options...] FQDN                                              \n".
         "     Otions:                                                                           \n".
         "        -d, --details     display a detailed verbose output                            \n".
         "        -c, --changelog   display the version history and exit                         \n".
         "        -v, --version     display the version number and exit                          \n".
         "        -h, --help        showing this information                                     \n\n".
         "     FQDN                 Fully Qualified Domain Name to be tested                     \n\n".
         "  In the default simple mode, the script returns 'all' and the return code 0 (success) \n".
         "  when any of DK/DKIM/ADSP policies require all messages to be signed. In all other    \n".
         "  cases (no policy defined, or allowing some messages without signatures), the return  \n".
         "  code is 1 (error), and the return string is 'some/undef'. In the detailed mode, the  \n".
         "  script returns DK/DKIM/ADSP policies as found in the DNS.                            \n";
my $changelog =
"  1.03 [Ivo Truxa] 2017/01/11                                                                   \n".
"       - added other forgotten policy strings (o=!, dkim=all, dkim=discardable)                 \n".
"  1.02 [Ivo Truxa] 2017/01/10                                                                   \n".
"       - return code 0 (enforced) or 1 (undefined) kept, but STDOUT prints a verbose result     \n".
"  1.01 [Ivo Truxa] 2017/01/09                                                                   \n".
"       - inversed output - 0: success = DKIM signing enforced                                   \n".
"       -                   1: error = policy undefined or signing not enforced                  \n".
"       - policy loop interrupted at the 1st hit, to save time, processing power, and bandwidth  \n".
"       - full policy loop for detailed verbose mode kept                                        \n".
"       - cosmetic formatting changes, editing of help texts                                     \n".
"  1.00 [Ivo Truxa] 2017/01/08                                                                   \n".
"       - this tool was written on FreeBSD, using Perl module Mail-DKIM-0.40 and Perl5-24        \n".
"       - at this date, it was not tested on other platforms or with other versions of Perl      \n";

use strict;
use warnings;
use Getopt::Long;
use Mail::DKIM::Verifier;
use Mail::DKIM::Policy;

$|=1;

# -----------------------------------------------------------------------
sub myend($) {print shift(),"\n"; exit 1;}
# -----------------------------------------------------------------------
my %opts;
GetOptions( \%opts, 'details', 'version', 'help', 'changelog');
$opts{'help'}      and myend("$u1  \n$u2  \n");
$opts{'changelog'} and myend("$u1  \n$changelog  \n");
$opts{'version'}   and myend("dkim_policy v$version ($verdate)  \n");
# -----------------------------------------------------------------------


my $domain = $ARGV[0];
my $result = 1;

if ($domain)
{
  # FQDN syntax checking of the domain string could be added here

  my $dkim  = Mail::DKIM::Verifier->new();
  # default result is 1 (error / no dkim enforcement)

  # feed a dummy message to the DKIM verifier object
  # passing so the queried domain name in the From field
  $dkim->PRINT("From: dkim\@".$domain."\r\n");
  $dkim->PRINT("To: dkim\@dkim.org\r\n\r\n");

  if ($opts{'details'}) {print $u1."\nDKIM policies for ".$domain.":\n";}

  # get all policies that apply to a verified message
  foreach my $policy ($dkim->policies)
  {
      if ($opts{'details'})
        {print $policy->name() . ": " . $policy->as_string() . "\n";}
      if ($policy->as_string() =~ "o=-|o=!|dkim=all|dkim=discardable")
        {
         $result = 0;
         if (!$opts{'details'}) {last;}
        }
  }
  $dkim->CLOSE;
  if ($opts{'details'})
       {print "\nResult: ";
        print $result ? "DKIM undefined or unenforced\n\n" : "DKIM enforced\n\n";}
  else {print $result ? "some/undef " : "all ";}
}
 else
{
  myend("$u1\n>>> ERROR: No domain name given! <<<\n\n$u2");
}

exit $result;
